bindings/python/triplestore/triplestore.py

```diff
@@ -36,23 +38,154 @@ class UniquenessError(TriplestoreError):
+            url = triplestore_url if triplestore_url else iri
+            triplestore = Triplestore("rdflib", base_iri=iri)
+            triplestore.parse(url)
+
+        self._cache = {} if cachemode != Namespace.NO_CACHE else None
+        #
+        # FIXME:
+        # Change this to only assigning the triplestore if cachemode is
+        # ONLY_CACHE when we figure out a good way to pre-populate the
+        # cache with IRIs from the triplestore.
+        #
+        #self._triplestore = (
+        #    triplestore if cachemode != Namespace.ONLY_CACHE else None
+        #)
+        self._triplestore = triplestore if need_triplestore else None
+
+        if cachemode != Namespace.NO_CACHE:
+            self._update_cache(triplestore)
+
+    def _update_cache(self, triplestore=None):
+        """Update the internal cache from `triplestore`."""
+        if not triplestore:
+            triplestore = self._triplestore
+        if not triplestore:
+            raise NamespaceError(
+                "`triplestore` argument needed for updating the cache"
+            )
+        if self._cache is None:
+            self._cache = {}
+
+        # Add (label, full_iri) pairs to cache
+        for la in reversed(self._label_annotations):
+            self._cache.update(
+                (o, s) for s, o in triplestore.subject_objects(la)
+                if s.startswith(self._iri)
+            )
+
+        # Add (name, full_iri) pairs to cache
+        # Currently we only check concepts that defines RDFS.isDefinedBy
+        # relations.
+        # Is there an efficient way to loop over all IRIs in this namespace?
+        n = len(self._iri)
+        self._cache.update(
+            (s[n:], s) for s in triplestore.subjects(
+                RDFS.isDefinedBy, self._iri)
+            if s.startswith(self._iri)
+        )
 
     def __getattr__(self, name):
-        return self.uri + name
+        if self._cache is not None and name in self._cache:
+            return self._cache[name]
+
+        if self._triplestore:
+
+            # Check if ``iri = self._iri + name`` is in the triplestore.
+            # If so, add it to the cache.
+            # We only need to check that generator returned by
+            # `self._triplestore.predicate_objects(iri)` is non-empty.
+            iri = self._iri + name
+            g = self._triplestore.predicate_objects(iri)
+            try:
+                g.__next__()
+            except StopIteration:
+                pass
+            else:
+                if self._cache is not None:
+                    self._cache[name] = iri
+                return iri
+
+            # Check for label annotations matching `name`.
+            for la in self._label_annotations:
+                for s, o in self._triplestore.subject_objects(la):
+                    if name == o:
+                        if self._cache is not None:
+                            self._cache[name]: s
```

Comment:
Did you mean to declare an assignment?

` self._cache[name] = s`

Reply:
Yes, that was the intention. Well spottet! 

I see now that there is another issue with this code as well. We only want to cache and return `s` if it belongs to the current namespace. Hence, I updated line 168 to
```python
    if o == name and s.startswith(self._iri):
```